File: open_mahjong_server/server/game_calculation/classical/classical_tingpai_check.py
from typing import List, Set
from time import time


class PlayerTiles:
    def __init__(self, tiles_list: List[int], combination_list: List[str], complete_step: int):
        self.hand_tiles = sorted(tiles_list)
        self.combination_list = combination_list
        self.complete_step = complete_step

    def __deepcopy__(self, memo):
        return PlayerTiles(self.hand_tiles[:], self.combination_list[:], self.complete_step)


class Classical_Tingpai_Check:
    def __init__(self):
        self.waiting_tiles: Set[int] = set()

    def check_waiting_tiles(self, player_tiles: PlayerTiles):
        self.waiting_tiles.clear()
        self.normal_check(player_tiles)
        return self.waiting_tiles

    def normal_check(self, player_tiles: PlayerTiles):
        if not self.normal_check_block(player_tiles):
            return

        all_list = self.normal_check_traverse_quetou(player_tiles)
        end_list = []
        while all_list:
            temp_list = all_list.pop()
            self.normal_check_traverse_kezi(temp_list, all_list)
            self.normal_check_traverse_dazi(temp_list, all_list)
            if temp_list.complete_step >= 11:
                end_list.append(temp_list)

        for i in end_list:
            if len(i.hand_tiles) == 1:
                self.waiting_tiles.add(i.hand_tiles[0])
            elif len(i.hand_tiles) == 2:
                if i.hand_tiles[0] == i.hand_tiles[1]:
                    self.waiting_tiles.add(i.hand_tiles[0])
                elif i.hand_tiles[0] == i.hand_tiles[1] - 1:
                    if i.hand_tiles[0] - 1 < 40:
                        self.waiting_tiles.add(i.hand_tiles[0] - 1)
                    if i.hand_tiles[0] + 2 < 40:
                        self.waiting_tiles.add(i.hand_tiles[0] + 2)
                elif i.hand_tiles[0] == i.hand_tiles[1] - 2:
                    if i.hand_tiles[0] + 1 < 40:
                        self.waiting_tiles.add(i.hand_tiles[0] + 1)

    def normal_check_block(self, player_tiles: PlayerTiles):
        if not player_tiles.hand_tiles:
            return True
        block_count = len(player_tiles.combination_list)
        tile_id_pointer = player_tiles.hand_tiles[0]
        for tile_id in player_tiles.hand_tiles:
            if tile_id == tile_id_pointer or tile_id == tile_id_pointer + 1:
                pass
            else:
                block_count += 1
            tile_id_pointer = tile_id
        return block_count <= 6

    def normal_check_traverse_quetou(self, player_tiles: PlayerTiles):
        all_list = []
        quetou_id_pointer = 0
        for tile_id in player_tiles.hand_tiles:
            if player_tiles.hand_tiles.count(tile_id) >= 2 and tile_id != quetou_id_pointer:
                temp_list = player_tiles.__deepcopy__(None)
                temp_list.hand_tiles.remove(tile_id)
                temp_list.hand_tiles.remove(tile_id)
                temp_list.complete_step += 2
                temp_list.combination_list.append(f"q{tile_id}")
                all_list.append(temp_list)
                quetou_id_pointer = tile_id
        all_list.append(player_tiles.__deepcopy__(None))
        return all_list

    def normal_check_traverse_kezi(self, player_tiles: PlayerTiles, all_list):
        same_tile_id = 0
        for tile_id in player_tiles.hand_tiles:
            if player_tiles.hand_tiles.count(tile_id) >= 3 and tile_id != same_tile_id:
                temp_list = player_tiles.__deepcopy__(None)
                temp_list.hand_tiles.remove(tile_id)
                temp_list.hand_tiles.remove(tile_id)
                temp_list.hand_tiles.remove(tile_id)
                temp_list.complete_step += 3
                temp_list.combination_list.append(f"k{tile_id}")
                all_list.append(temp_list)
                same_tile_id = tile_id

    def normal_check_traverse_dazi(self, player_tiles: PlayerTiles, all_list):
        same_tile_id = 0
        for tile_id in player_tiles.hand_tiles:
            if tile_id <= 40:
                if tile_id + 1 in player_tiles.hand_tiles and tile_id + 2 in player_tiles.hand_tiles and tile_id != same_tile_id:
                    temp_list = player_tiles.__deepcopy__(None)
                    temp_list.hand_tiles.remove(tile_id)
                    temp_list.hand_tiles.remove(tile_id + 1)
                    temp_list.hand_tiles.remove(tile_id + 2)
                    temp_list.complete_step += 3
                    temp_list.combination_list.append(f"s{tile_id + 1}")
                    all_list.append(temp_list)
                    same_tile_id = tile_id

    def tingpai_check(self, hand_tile_list: List[int], combination_list: List[str]):
        test_tiles = PlayerTiles(hand_tile_list.copy(), combination_list.copy(), len(combination_list) * 3)
        self.check_waiting_tiles(test_tiles)
        self.waiting_tiles = {i for i in self.waiting_tiles if i not in {10, 20, 30, 40}}
        return self.waiting_tiles.copy()


if __name__ == "__main__":
    test_hand_tiles = [11, 12, 13, 21, 22, 23, 32, 33, 35, 35, 37, 38, 39]
    test_combination_list = []
    checker = Classical_Tingpai_Check()
    t0 = time()
    result = checker.tingpai_check(test_hand_tiles, test_combination_list)
    t1 = time()
    print("Classical Tingpai Check")
    print("手牌:", test_hand_tiles)
    print("副露:", test_combination_list)
    print("听牌:", sorted(result))
    print("耗时:", t1 - t0)

File: open_mahjong_unity/Assets/Scripts/GameScene/Calculation/CalculationScript/Classical/ClassicalCalc/ClassicalTingpaiCheck.cs
using System.Collections.Generic;

namespace Classical {
    /// <summary>
    /// 古典麻将听牌检测。
    /// </summary>
    public class ClassicalTingpaiCheck {
        private readonly HashSet<int> _waitingTiles = new HashSet<int>();
        private static readonly HashSet<int> InvalidTiles = new HashSet<int> { 10, 20, 30, 40 };

        /// <summary>
        /// 外部调用入口：传入手牌（不含和牌张，即 13 / 10 / 7 / 4 / 1 张）与副露列表，返回听牌集合。
        /// </summary>
        public HashSet<int> TingpaiCheck(List<int> handTileList, List<string> combinationList) {
            int completeStep = combinationList.Count * 3;
            var pt = new ClassicalPlayerTiles(handTileList, combinationList, completeStep);
            CheckWaitingTiles(pt);
            _waitingTiles.RemoveWhere(t => InvalidTiles.Contains(t));
            return new HashSet<int>(_waitingTiles);
        }

        /// <summary>
        /// 静态便捷方法，供 External 层调用。
        /// </summary>
        public static HashSet<int> TingpaiCheckStatic(List<int> handTileList, List<string> combinationList, bool debug = false) {
            var checker = new ClassicalTingpaiCheck();
            return checker.TingpaiCheck(handTileList, combinationList);
        }

        private void CheckWaitingTiles(ClassicalPlayerTiles pt) {
            _waitingTiles.Clear();
            NormalCheck(pt);
        }

        private void NormalCheck(ClassicalPlayerTiles pt) {
            if (!BlockCheck(pt)) return;

            var allList = TraverseQuetou(pt);
            var endList = new List<ClassicalPlayerTiles>();
            while (allList.Count > 0) {
                var cur = allList[allList.Count - 1];
                allList.RemoveAt(allList.Count - 1);
                TraverseKezi(cur, allList);
                TraverseDazi(cur, allList);
                if (cur.CompleteStep >= 11)
                    endList.Add(cur);
            }

            foreach (var item in endList) {
                if (item.HandTiles.Count == 1) {
                    _waitingTiles.Add(item.HandTiles[0]);
                } else if (item.HandTiles.Count == 2) {
                    int a = item.HandTiles[0], b = item.HandTiles[1];
                    if (a == b) {
                        _waitingTiles.Add(a);
                    } else if (a == b - 1) {
                        if (a - 1 > 0 && a - 1 < 40) _waitingTiles.Add(a - 1);
                        if (a + 2 > 0 && a + 2 < 40) _waitingTiles.Add(a + 2);
                    } else if (a == b - 2) {
                        if (a + 1 > 0 && a + 1 < 40) _waitingTiles.Add(a + 1);
                    }
                }
            }
        }

        private static bool BlockCheck(ClassicalPlayerTiles pt) {
            if (pt.HandTiles.Count == 0) return true;
            int blockCount = pt.CombinationList.Count;
            int pointer = pt.HandTiles[0];
            foreach (int t in pt.HandTiles) {
                if (t != pointer && t != pointer + 1)
                    blockCount++;
                pointer = t;
            }
            return blockCount <= 6;
        }

        private static List<ClassicalPlayerTiles> TraverseQuetou(ClassicalPlayerTiles pt) {
            var result = new List<ClassicalPlayerTiles>();
            int prevId = 0;
            foreach (int t in pt.HandTiles) {
                if (CountInList(pt.HandTiles, t) >= 2 && t != prevId) {
                    var copy = pt.DeepCopy();
                    copy.HandTiles.Remove(t);
                    copy.HandTiles.Remove(t);
                    copy.CompleteStep += 2;
                    copy.CombinationList.Add($"q{t}");
                    result.Add(copy);
                    prevId = t;
                }
            }
            result.Add(pt.DeepCopy());
            return result;
        }

        private static void TraverseKezi(ClassicalPlayerTiles pt, List<ClassicalPlayerTiles> allList) {
            int prevId = 0;
            foreach (int t in pt.HandTiles) {
                if (CountInList(pt.HandTiles, t) >= 3 && t != prevId) {
                    var copy = pt.DeepCopy();
                    copy.HandTiles.Remove(t);
                    copy.HandTiles.Remove(t);
                    copy.HandTiles.Remove(t);
                    copy.CompleteStep += 3;
                    copy.CombinationList.Add($"k{t}");
                    allList.Add(copy);
                    prevId = t;
                }
            }
        }

        private static void TraverseDazi(ClassicalPlayerTiles pt, List<ClassicalPlayerTiles> allList) {
            int prevId = 0;
            foreach (int t in pt.HandTiles) {
                if (t <= 40 && pt.HandTiles.Contains(t + 1) && pt.HandTiles.Contains(t + 2) && t != prevId) {
                    var copy = pt.DeepCopy();
                    copy.HandTiles.Remove(t);
                    copy.HandTiles.Remove(t + 1);
                    copy.HandTiles.Remove(t + 2);
                    copy.CompleteStep += 3;
                    copy.CombinationList.Add($"s{t + 1}");
                    allList.Add(copy);
                    prevId = t;
                }
            }
        }

        private static int CountInList(List<int> list, int value) {
            int count = 0;
            for (int i = 0; i < list.Count; i++) {
                if (list[i] == value) count++;
            }
            return count;
        }
    }
}
